network/p2p/libp2pNode.go

```diff
@@ -490,28 +496,66 @@ func (n *Node) IsConnected(identity flow.Identity) (bool, error) {
 	return isConnected, nil
 }
 
-// bootstrapLibP2PHost creates and starts a libp2p host as well as a pubsub component for it, and returns all in a
-// libP2PHostWrapper.
-// In case `allowList` is true, it also creates and embeds a connection gater in the returned libP2PHostWrapper, which
-// whitelists the `allowListAddres` nodes.
-func bootstrapLibP2PHost(ctx context.Context,
-	logger zerolog.Logger,
+// WithDefaultLibP2PHost returns a NodeOption that instantiates a libp2p host with the default values for the libp2p
+// transport, ping etc.
+// If 'allowList' is true, a connection gater is also added to the host
+func WithDefaultLibP2PHost(
 	address string,
 	conMgr *ConnManager,
-	key crypto.PrivKey,
-	allowList bool,
-	psOption ...pubsub.Option) (host.Host, *connGater, *pubsub.PubSub, error) {
+	key fcrypto.PrivateKey,
+	allowList bool) NodeOption {
+
+	return func(node *Node) error {
 
-	var connGater *connGater
+		options, err := DefaultLibP2POptions(address, key, true)
+		if err != nil {
+			return err
+		}
+
+		if conMgr != nil {
+			options = append(options, libp2p.ConnectionManager(conMgr)) // set the connection manager
+			node.connMgr = conMgr
+		}
+
+		// if allowlisting is enabled, create a connection gator with allowListAddrs
+		if allowList {
+			// create a connection gater
+			connGater := newConnGater(node.logger)
+
+			// provide the connection gater as an option to libp2p
+			options = append(options, libp2p.ConnectionGater(connGater))
+
+			node.connGater = connGater
+		}
+
+		// create the libp2p host
+		libP2PHost, err := libp2p.New(node.ctx, options...)
```

Comment:
Generally I think storing a context inside a struct is considered an antipattern. See the last two paragraphs of https://github.com/dapperlabs/flow-go/issues/5720 for reasons why, and links to a blog post from the Go team.

I think we should avoid storing `ctx` on the node, just so it can be referenced here. Instead, how about we change the type definition of `NodeOption` to the following:

```go
type NodeOption func(node *Node, ctx context.Context)
```

Then, the ctx can directly be passed in from the node constructor.

Reply:
On second thought, what would be better is if we can move any logic that creates subcomponents out of these `NodeOption` functions and into the constructor function itself so that the `ctx` can be passed directly.

We should use the `NodeOption` functions merely to pass the parameters that will be used to create these subcomponents. The actual creation of the component can happen in the constructor